### DISCUSSION TITLE
Manifold backend?

### DISCUSSION THREAD

**Discussion by pca006132** [geoffder/dometyl-keyboard#8] [General]
Just wondering what features you use with OpenSCAD are not yet implemented in manifold. I am thinking about playing with these generators but OpenSCAD is a bit too slow, even with the fast-csg thing in nightly. It would be nice if we can make the manifold backend compatible with this.

**Comment by geoffder** [comment]
Ah nice to hear that you're interested in my humble keyboard jank! 

I've just taken a quick look, and I most of the remaining uses of `Scad.hull` could be replaced easily with one of`OCADml`'s hulling [functions](https://ocadml.github.io/OCADml/OCADml/OCADml/Mesh/index.html#val-hull), and the one in `Hand` that would be more effort (but not much more to be honest) could fall back on [`OManifold.Manifold.hull`](https://github.com/OCADml/OManifold/blob/9ad49f0ebce72434bfc326e138b2e2d0800e3347/lib/manifold.ml#L376). So, the only thing really that isn't covered by Manifold currently that is used (but not too much, the hotswap sockets in key modules being the main ones that come to mind), are 2d boolean operations. I've recently put together Clipper2 bindings in OCaml that could cover those 2d booleans.

For accessibility reasons, I think OpenSCAD will always be the default backend for this generator, but it might be interesting to try functorizing over the backend (the interface of OManifold is quite similar to OSCADml already). For reference, my recent [Clipper2 binding library](https://github.com/geoffder/ocaml-clipper2/blob/main/lib/clipper.ml) is an example of using OCamls module functors to support user provided vector and polygon types. Experimenting with swapping over to Manifold in place by replacing `Scad` with `Manifold` would take care of a lot of it if Manifold had a matching 2d subsystem and I was able to abstract over it in a similar way as I have with OSCADml (see the [Scad GADT](https://github.com/OCADml/OSCADml/blob/7b00751b96ffb096f506b3344355f30fa7e2be51/lib/scad.mli#L18), if it makes any sense to non-ocamler eyes).

By the way, have you tried out this generator in particular and found OpenSCAD to still be too sluggish with `fast-csg` etc enabled? With hotswap sockets off (empty keyholes), and segment counts turned down on the walls etc, rendering (F6) is less than 10s, and high quality with hotswap is less than 80s in my check just now. Preview is not smooth, but with render times so short I don't find that so bad. One recent realization I had was the bad file loading and preview performance issues I had were due to the editor attempting to highlight and wrap the large (several MB and beyond) generated files. Generating two files and including the one in the other lead to huge gains.

**Reply by pca006132** [reply]
No I haven't really tried it yet. The last time I use OpenSCAD is already half a year ago and I switched over to manifold as my design would compile for a few minutes with fast-csg while manifold would take about a second. I'm also writing a manuform keyboard generator which can export the model really quickly, but I found thumb cluster really hard to model, so I wanted to try some existing solutions and see if they fit my need :). I will try OpenSCAD one first, but please let me know if you encountered any problem with using Manifold as backend!

**Reply by geoffder** [reply]
Will do! Not sure when it'll happen, but you've got me curious to experiment with it at some point.
